### PR TITLE
Improve performance of updatecheck data aggregation

### PR DESCRIPTION
**Pull request by efritz**
This PR improves the performance of the updatecheck client. Differences to current approach:

- We use @tsenart's approach from https://github.com/sourcegraph/sourcegraph/issues/10612 of doing a single query pass over the event logs table. This returns the total number of events, unique users, and latencies of a set of whitelisted events for the current day, week, and month (one each).
- We massage the aggregated events from the DB into a {CodeIntel/Search}UsageStatisticsObject to be serialized by the update check handler instead of doing 27+ individual queries to populate these objects. Keeping the update check structs the same means no other changes are necessary to the payload formats.
- We return data only for time periods equal to `DATE_TRUNC(now, 'month'|'week'|'day')`. In order to do multiple time periods, we would need to alter the query to bring back a richer set of data.

A next step will be to remove the code that's effectively dead after this change is validated and merged.

Closes https://github.com/sourcegraph/sourcegraph/issues/10612.

### CODE REVIEW

File: cmd/frontend/db/event_logs.go [hunk 1]
```diff
@@ -600,3 +601,121 @@ FROM event_logs
...
+      -- It's faster to group by an int rather than text, so we convert
+      -- the anonymous_user_id to an int, rather than the user_id to text.
+      THEN ('x'||substr(md5(anonymous_user_id), 1, 8))::bit(32)::int
+      ELSE user_id
+    END AS user_id,
+    DATE(TIMEZONE('UTC', timestamp)) AS date
+  FROM event_logs
+  WHERE timestamp >= DATE_TRUNC('month', NOW())
+) q
+WHERE name IN (%s) GROUP BY name
```
Comment (tsenart) [inline cmd/frontend/db/event_logs.go:720]: I think we can filter by the name in the sub-query already.

```sql
WHERE timestamp >= DATE_TRUNC('month', NOW()) AND name in (%s)
```

File: cmd/frontend/internal/app/pkg/updatecheck/client.go [hunk 2]
```diff
@@ -245,13 +233,9 @@ func updateBody(ctx context.Context) (io.Reader, error) {
 		if err != nil {
 			logFunc("updatecheck.getAndMarshalCampaignsUsageJSON failed", "error", err)
 		}
-		r.CodeIntelUsage, err = getAndMarshalCodeIntelUsageJSON(ctx)
-		if err != nil {
-			logFunc("updatecheck.getAndMarshalCodeIntelUsageJSON failed", "error", err)
-		}
-		r.SearchUsage, err = getAndMarshalSearchUsageJSON(ctx)
+		r.CodeIntelUsage, r.SearchUsage, err = getAndMarshalAggregatedUsageJSON(ctx)
```
Comment (tsenart) [inline cmd/frontend/internal/app/pkg/updatecheck/client.go:236]: I'm concerned that leaving `getAndMarshalSiteActivityJSON` as it is will keep the run-time of this method too high still. Can we use the output of the query for those statistics as well?

Comment (efritz) [reply cmd/frontend/internal/app/pkg/updatecheck/client.go:236]: Yes, sorry I missed it. Will update now.

Comment (efritz) [reply cmd/frontend/internal/app/pkg/updatecheck/client.go:236]: How do we do this with the same query, though? It needs unique users over all events, but the group by is pretty aggressive.

File: cmd/frontend/db/event_logs.go (hunk 1 above)
Comment (efritz) [reply cmd/frontend/db/event_logs.go:720]: Updated.

File: cmd/frontend/internal/app/pkg/updatecheck/client.go (hunk 2 above)
Comment (tsenart) [reply cmd/frontend/internal/app/pkg/updatecheck/client.go:236]: Ah, yes, for `getAndMarshalSiteActivityJSON` we might need a separate query. Can you adapt this one for those needs somehow? I can also take a stab at that next week, as to not block this work. Then, we should probably run these two queries concurrently.

File: cmd/frontend/db/event_logs.go [hunk 3]
```diff
@@ -600,3 +601,121 @@ FROM event_logs
...
+	"codeintel.searchReferences",
+	"search.latencies.literal",
+	"search.latencies.regexp",
+	"search.latencies.structural",
+	"search.latencies.file",
+	"search.latencies.repo",
+	"search.latencies.diff",
+	"search.latencies.commit",
+	"search.latencies.symbol",
+}
```
Comment (tsenart) [inline cmd/frontend/db/event_logs.go:666]: Can we put these directly in the SQL query?

File: cmd/frontend/internal/app/pkg/updatecheck/client.go (hunk 2 above)
Comment (efritz) [reply cmd/frontend/internal/app/pkg/updatecheck/client.go:236]: Sure. I'll work on optimizing that query, but in a separate PR. This query seemed to be fine before we added all the codeintel/search queries, which I think is the largest factor.

File: cmd/frontend/db/event_logs.go (hunk 3 above)
Comment (efritz) [reply cmd/frontend/db/event_logs.go:666]: Sure.

**Comment by efritz** [comment]
@tsenart Small update to fix when weeks start. Please take another look if you can.

File: cmd/frontend/db/event_logs.go [hunk 4]
```diff
@@ -600,3 +601,112 @@ FROM event_logs
...
+FROM (
+  -- This sub-query is here to avoid re-doing this work above on each aggregation.
+  SELECT
+    name,
+    -- Postgres 9.6 needs to go from text to integer (i.e. can't go directly to integer)
+    (argument->'durationMs')::text::integer as latency,
+    CASE WHEN user_id = 0
+      -- It's faster to group by an int rather than text, so we convert
+      -- the anonymous_user_id to an int, rather than the user_id to text.
+      THEN ('x'||substr(md5(anonymous_user_id), 1, 8))::bit(32)::int
```
Comment (unknwon) [inline cmd/frontend/db/event_logs.go:689]: 👍 TIL

**Comment by tsenart** [comment]
Pushed a commit that simplifies the query a bit. PTAL.

File: cmd/frontend/db/event_logs.go [hunk 5]
```diff
@@ -600,3 +601,105 @@ FROM event_logs
...
+			&event.Month,
+			&event.Week,
+			&event.Day,
+			&event.TotalMonth,
+			&event.TotalWeek,
+			&event.TotalDay,
+			&event.UniquesMonth,
+			&event.UniquesWeek,
+			&event.UniquesDay,
+			pq.Array(&event.LatenciesMonth),
```
Comment (tsenart) [inline cmd/frontend/db/event_logs.go:626]: One thing I noticed during testing is that sometimes we can have NULL returned for these arrays when, for instance, there hasn't been a single event of the kind in the day the query runs. Does pq.Array handle that?

Comment (efritz) [reply cmd/frontend/db/event_logs.go:626]: I'll add a test to make sure it does.

**Comment by efritz** [comment]
@dadlerj @tsenart @unknwon I have added a new query to replace site usage statistics as well. Please take another look.

File: cmd/frontend/db/event_logs.go [hunk 6]
```diff
@@ -600,3 +601,238 @@ FROM event_logs
 GROUP BY 1, 2
 ORDER BY 1 DESC, 2 ASC;
 `
+
+// SiteUsage calculates AggregatedEvent for each every unique event type.
```
Comment (tsenart) [inline cmd/frontend/db/event_logs.go:605]: This comment is weird.

**Comment by tsenart** [comment]
@efritz: The query you used for site summary isn't the most up to date I left in the issue. Can you update it please?

**Comment by tsenart** [comment]
Here's the link: https://github.com/sourcegraph/sourcegraph/issues/10612#issuecomment-630190551

File: cmd/frontend/db/event_logs_test.go [hunk 7]
```diff
@@ -254,6 +254,272 @@ func TestEventLogs_UsersUsageCounts(t *testing.T) {
...
+	dbtesting.SetupGlobalTestDB(t)
+	ctx := context.Background()
+
+	// This unix timestamp is equivalent to `Friday, May 15, 2020 10:30:00 PM GMT` and is set to
+	// be a consistent value so that the tests don't fail when someone runs it at some particular
+	// time that falls too near the edge of a week.
+	now := time.Unix(1589581800, 0).UTC()
+
+	days := map[time.Time]struct {
+		users   []uint32
```
Comment (tsenart) [inline cmd/frontend/db/event_logs_test.go:270]: He, this was actually the bug in the first version of the query — it didn't consider anonymous users, which we should test here.

File: cmd/frontend/types/types.go [hunk 8]
```diff
@@ -302,6 +302,49 @@ type SearchEventLatencies struct {
 	P99 float64
 }
 
+// AggregatedEvent is an alternate view of SiteUsageStatistics which is
```
Comment (tsenart) [inline cmd/frontend/types/types.go:305]: Comment fix please.

**Comment by tsenart** [comment]
@efritz: It would be great to get this in to the 3.16 release. Can we merge this, test that it works on sourcegraph.com, and if so, cherry-pick it into the 3.16 release branch? If we don't, we'll have to issue a patch release for large customers anyway... cc @uwedeportivo

**Comment by tsenart** [comment]
@dadlerj: We merged this to test in sourcegraph.com in an attempt to include this in the 3.16 release. Please review post-merge today once you come online. I don't believe you'd have substantial feedback, but we wanna be sure.

**Comment by dadlerj** [comment]
Thanks @tsenart and @efritz . Sorry I didn't review, I was under the impression we weren't planning to merge pre-3.16!

Like you said, I won't have much to say here — as long as the output is still the same, I'll be happy. The only risk is if something DOES appear broken today, we'll need to revert this for 3.16.

We should know soon whether this causes any problems on dotcom :)

**Comment by tsenart** [comment]
@efritz: This graph seems to indicate that we still have a 10s timeout on that function. I think we'd need to fix that and perhaps find a way to test more directly? https://sourcegraph.com/-/debug/grafana/explore?orgId=1&left=%5B%22now-7d%22,%22now%22,%22Prometheus%22,%7B%22expr%22:%22histogram_quantile(0.95,%20sum(rate(src_updatecheck_duration_seconds_bucket%5B5m%5D))%20by%20(le))%22%7D,%7B%22mode%22:%22Metrics%22%7D,%7B%22ui%22:%5Btrue,true,true,%22none%22%5D%7D%5D

**Pull request merged**